File: DevExpress-Demo/CS/WindowsUIDemo.Wpf/ViewModels/NotificationsViewModel.cs
using DevExpress.Mvvm;
using DevExpress.Mvvm.DataAnnotations;
using DevExpress.Mvvm.POCO;
using DevExpress.Mvvm.UI;
using DevExpress.Mvvm.UI.Interactivity;
using DevExpress.Xpf.DemoBase.Helpers;
using System;
using System.Threading.Tasks;
using System.Windows.Media;
using System.Windows.Media.Imaging;

namespace WindowsUIDemo {
    public class NotificationsViewModel {
        public virtual bool ShowImage { get; set; }
        public virtual NotificationTemplate Template { get; set; }
        public virtual INotificationService NotificationService { get { return null; } }
        public virtual ILogService LogService { get { return null; } }

        public string AppModelID { get; private set; }
        public int PredefinedNotificationId { get; set; }

        protected NotificationsViewModel() {
            ShowImage = true;
            CustomNotificationWidth = 380;
            CustomNotificationHeight = 100;
            CustomNotificationBackground = Color.FromRgb(0xf7, 0x94, 0x1e);
            PredefinedNotificationId = 100;
        }

        public void ShowCustom() {
            CustomToastViewModel viewModel = ViewModelSource.Create(() => new CustomToastViewModel {
                Width = CustomNotificationWidth,
                Height = CustomNotificationHeight,
                Text = CustomNotificationText,
                Background = CustomNotificationBackgroundBrush
            });
            INotification notification = NotificationService.CreateCustomNotification(viewModel);
            Show(notification);
        }

        public void ShowPredefined() {
            ImageSource image = ShowImage ? new BitmapImage(new Uri("pack://application:,,,/WindowsUIDemo;component/Images/sun.png", UriKind.Absolute)) : null;
            string text1 = "Lorem ipsum dolor sit amet integer fringilla, dui eget ultrices cursus, justo tellus.";
            string text2 = "In ornare ante magna, eget volutpat mi bibendum a. Nam ut ullamcorper libero. Pellentesque habitant.";
            string text3 = "Quisque sapien odio, mollis tincidunt";
            INotification notification = NotificationService.CreatePredefinedNotification(text1, text2, text3, image, PredefinedNotificationId.ToString());
            PredefinedNotificationId++;
            Show(notification);
        }

        void Show(INotification notification) {
            LogService.LogLine("Showing...");
            notification.ShowAsync().ContinueWith(OnNotificationShown, TaskScheduler.FromCurrentSynchronizationContext());
        }

        void OnNotificationShown(Task<NotificationResult> task) {
            try {
                switch(task.Result) {
                    case NotificationResult.Activated:
                        LogService.LogLine("Activated");
                        break;
                    case NotificationResult.TimedOut:
                        LogService.LogLine("Timed out");
                        break;
                    case NotificationResult.UserCanceled:
                        LogService.LogLine("Canceled by user");
                        break;
                    case NotificationResult.Dropped:
                        LogService.LogLine("Dropped (the queue is full)");
                        break;
                }
            } catch(AggregateException e) {
                var actualException = e.InnerException.InnerException;
                var exceptionMessage = actualException != null ? actualException.Message : e.InnerException.Message;
                LogService.LogLine(exceptionMessage.ToUpper().Contains("803E0114")
                    ? "System settings prevent native notifications from being delivered. You can enable notifications in the Settings app: Settings | System | Notifications & actions."
                    : ("Error: " + exceptionMessage));
            }
            LogService.LogLine("");
        }

        protected void OnCustomNotificationBackgroundChanged() {
            CustomNotificationBackgroundBrush = new SolidColorBrush { Color = CustomNotificationBackground };
        }

        public virtual string CustomNotificationText { get; set; }
        public virtual double CustomNotificationWidth { get; set; }
        public virtual double CustomNotificationHeight { get; set; }
        public virtual Color CustomNotificationBackground { get; set; }
        public virtual SolidColorBrush CustomNotificationBackgroundBrush { get; set; }
    }

    [POCOViewModel]
    public class CustomToastViewModel {
        public virtual string Text { get; set; }
        public virtual double Height { get; set; }
        public virtual double Width { get; set; }
        public virtual SolidColorBrush Background { get; set; }
    }
    public class NotificationInfo {
        public NotificationTemplate Template { get; set; }
        public string Description { get; set; }
        public string Win10Description { get; set; }
    }
}

File: DevExpress-Demo/VB/WindowsUIDemo.Wpf/ViewModels/NotificationsViewModel.vb
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.DataAnnotations
Imports DevExpress.Mvvm.POCO
Imports DevExpress.Mvvm.UI
Imports DevExpress.Xpf.DemoBase.Helpers
Imports System
Imports System.Threading.Tasks
Imports System.Windows.Media
Imports System.Windows.Media.Imaging

Namespace WindowsUIDemo

    Public Class NotificationsViewModel

        Private _AppModelID As String

        Public Overridable Property ShowImage As Boolean

        Public Overridable Property Template As NotificationTemplate

        Public Overridable ReadOnly Property NotificationService As INotificationService
            Get
                Return Nothing
            End Get
        End Property

        Public Overridable ReadOnly Property LogService As ILogService
            Get
                Return Nothing
            End Get
        End Property

        Public Property AppModelID As String
            Get
                Return _AppModelID
            End Get

            Private Set(ByVal value As String)
                _AppModelID = value
            End Set
        End Property

        Public Property PredefinedNotificationId As Integer

        Protected Sub New()
            ShowImage = True
            CustomNotificationWidth = 380
            CustomNotificationHeight = 100
            CustomNotificationBackground = Color.FromRgb(&Hf7, &H94, &H1e)
            PredefinedNotificationId = 100
        End Sub

        Public Sub ShowCustom()
            Dim viewModel As CustomToastViewModel = ViewModelSource.Create(Function() New CustomToastViewModel With {.Width = CustomNotificationWidth, .Height = CustomNotificationHeight, .Text = CustomNotificationText, .Background = CustomNotificationBackgroundBrush})
            Dim notification As INotification = NotificationService.CreateCustomNotification(viewModel)
            Show(notification)
        End Sub

        Public Sub ShowPredefined()
            Dim image As ImageSource = If(ShowImage, New BitmapImage(New Uri("pack://application:,,,/WindowsUIDemo;component/Images/sun.png", UriKind.Absolute)), Nothing)
            Dim text1 As String = "Lorem ipsum dolor sit amet integer fringilla, dui eget ultrices cursus, justo tellus."
            Dim text2 As String = "In ornare ante magna, eget volutpat mi bibendum a. Nam ut ullamcorper libero. Pellentesque habitant."
            Dim text3 As String = "Quisque sapien odio, mollis tincidunt"
            Dim notification As INotification = NotificationService.CreatePredefinedNotification(text1, text2, text3, image, PredefinedNotificationId.ToString())
            PredefinedNotificationId += 1
            Show(notification)
        End Sub

        Private Sub Show(ByVal notification As INotification)
            LogService.LogLine("Showing...")
            notification.ShowAsync().ContinueWith(New Action(Of Task(Of NotificationResult))(AddressOf OnNotificationShown), TaskScheduler.FromCurrentSynchronizationContext())
        End Sub

        Private Sub OnNotificationShown(ByVal task As Task(Of NotificationResult))
            Try
                Select Case task.Result
                    Case NotificationResult.Activated
                        LogService.LogLine("Activated")
                    Case NotificationResult.TimedOut
                        LogService.LogLine("Timed out")
                    Case NotificationResult.UserCanceled
                        LogService.LogLine("Canceled by user")
                    Case NotificationResult.Dropped
                        LogService.LogLine("Dropped (the queue is full)")
                End Select
            Catch e As AggregateException
                Dim actualException = e.InnerException.InnerException
                Dim exceptionMessage = If(actualException IsNot Nothing, actualException.Message, e.InnerException.Message)
                LogService.LogLine(If(exceptionMessage.ToUpper().Contains("803E0114"), "System settings prevent native notifications from being delivered. You can enable notifications in the Settings app: Settings | System | Notifications & actions.", "Error: " & exceptionMessage))
            End Try

            LogService.LogLine("")
        End Sub

        Protected Sub OnCustomNotificationBackgroundChanged()
            CustomNotificationBackgroundBrush = New SolidColorBrush With {.Color = CustomNotificationBackground}
        End Sub

        Public Overridable Property CustomNotificationText As String

        Public Overridable Property CustomNotificationWidth As Double

        Public Overridable Property CustomNotificationHeight As Double

        Public Overridable Property CustomNotificationBackground As Color

        Public Overridable Property CustomNotificationBackgroundBrush As SolidColorBrush
    End Class

    <POCOViewModel>
    Public Class CustomToastViewModel

        Public Overridable Property Text As String

        Public Overridable Property Height As Double

        Public Overridable Property Width As Double

        Public Overridable Property Background As SolidColorBrush
    End Class

    Public Class NotificationInfo

        Public Property Template As NotificationTemplate

        Public Property Description As String

        Public Property Win10Description As String
    End Class
End Namespace
